### PR TITLE
Unpack fields inside CardanoTxOut to reduce memory footprint

### PR DESCRIPTION
**Pull request by jorisdral**

### CODE REVIEW

File: ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs [hunk 1]
```diff
@@ -72,13 +73,25 @@ instance CardanoHardForkConstraints c
 
   decodeCanonicalTxIn = CardanoTxIn <$> Core.fromEraCBOR @(ShelleyEra c)
 
+-- Unpacking the fields of the era-specific TxOuts could save a chunk of memory.
+-- However, unpacking of sum types is only possible on @ghc-9.6.1@ and later, so
+-- before @ghc-9.6.1@ we only unpack the TxOuts for ears before Alonzo.
```
Comment (amesgen) [inline ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs:78]: :ear:s before Alonzo

File: ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs [hunk 2]
```diff
@@ -72,13 +73,25 @@ instance CardanoHardForkConstraints c
...
 data CardanoTxOut c =
-    ShelleyTxOut !(Value (LedgerState (ShelleyBlock (TPraos c) (ShelleyEra c))))
-  | AllegraTxOut !(Value (LedgerState (ShelleyBlock (TPraos c) (AllegraEra c))))
-  | MaryTxOut    !(Value (LedgerState (ShelleyBlock (TPraos c) (MaryEra c))))
+#if MIN_VERSION_GLASGOW_HASKELL(9,6,1,0)
```
Comment (jasagredo) [inline ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs:80]: We usually use `__GLASGOW_HASKELL__ > 906`, couldn't we use it also here or is it due to the patch version? I don't think `__GLASGOW_HASKELL__` has a minor field

File: ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs [hunk 3]
```diff
@@ -72,13 +73,25 @@ instance CardanoHardForkConstraints c
...
+  | AllegraTxOut {-# UNPACK #-} !(Value (LedgerState (ShelleyBlock (TPraos c) (AllegraEra c))))
+  | MaryTxOut    {-# UNPACK #-} !(Value (LedgerState (ShelleyBlock (TPraos c) (MaryEra c))))
+  | AlonzoTxOut  {-# UNPACK #-} !(Value (LedgerState (ShelleyBlock (TPraos c) (AlonzoEra c))))
+  | BabbageTxOut {-# UNPACK #-} !(Value (LedgerState (ShelleyBlock (Praos c) (BabbageEra c))))
+  | ConwayTxOut  {-# UNPACK #-} !(Value (LedgerState (ShelleyBlock (Praos c) (ConwayEra c))))
+#else
+    ShelleyTxOut {-# UNPACK #-} !(Value (LedgerState (ShelleyBlock (TPraos c) (ShelleyEra c))))
+  | AllegraTxOut {-# UNPACK #-} !(Value (LedgerState (ShelleyBlock (TPraos c) (AllegraEra c))))
+  | MaryTxOut    {-# UNPACK #-} !(Value (LedgerState (ShelleyBlock (TPraos c) (MaryEra c))))
   | AlonzoTxOut  !(Value (LedgerState (ShelleyBlock (TPraos c) (AlonzoEra c))))
```
Comment (jasagredo) [inline ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs:91]: I don't understand why we can do it for before alonzo and not for alonzo? if unpacking in sum types was not supported, why can it work for earlier eras?

Comment (jorisdral) [reply ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs:91]: https://github.com/IntersectMBO/cardano-ledger/blob/e47a836b45c6ad7384dc9ba270c752284a947a92/eras/shelley/impl/src/Cardano/Ledger/Shelley/TxOut.hs#L59-L62

```haskell
data ShelleyTxOut era = TxOutCompact
  { txOutCompactAddr :: {-# UNPACK #-} !(CompactAddr (EraCrypto era))
  , txOutCompactValue :: !(CompactForm (Value era))
  }
```
```haskell
type TxOut (MaryEra c) = ShelleyTxOut (MaryEra c)
```
TxOuts in earlier eras aren't sums. It's only starting from `Alonzo` that they become sums
```haskell
data AlonzoTxOut era
  = TxOutCompact'
      {-# UNPACK #-} !(CompactAddr (EraCrypto era))
      !(CompactForm (Value era))
  | TxOutCompactDH'
      {-# UNPACK #-} !(CompactAddr (EraCrypto era))
      !(CompactForm (Value era))
      !(DataHash (EraCrypto era))
  | TxOut_AddrHash28_AdaOnly
      !(Credential 'Staking (EraCrypto era))
      {-# UNPACK #-} !Addr28Extra
      {-# UNPACK #-} !(CompactForm Coin) -- Ada value
  | TxOut_AddrHash28_AdaOnly_DataHash32
      !(Credential 'Staking (EraCrypto era))
      {-# UNPACK #-} !Addr28Extra
      {-# UNPACK #-} !(CompactForm Coin) -- Ada value
      {-# UNPACK #-} !DataHash32
```

File: ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs (hunk 2 above)
Comment (jorisdral) [reply ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs:80]: The `z` in `MIN_VERSION_GLASGOW_HASKELL(x,y,z,z')` is important here. It's only from `ghc-9.6.1` onwards that we have this feature. `__GLASGOW_HASKELL__ > xyy` wouldn't let us put a bound on the `z`

Sources:
https://downloads.haskell.org/ghc/latest/docs/users_guide/phases.html?highlight=cpp#standard-cpp-macros
https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/pragmas.html#unpack-pragma

File: ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs (hunk 3 above)
Comment (jasagredo) [reply ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs:94]: Aha, I see :+1:

File: ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs (hunk 2 above)
Comment (jasagredo) [reply ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs:83]: Alright, clear now, thanks!

**Review by jasagredo — APPROVED** [review]

**Pull request merged**